### ISSUE TITLE
`isinstance(x, type)` incorrectly narrows `Union[type, Callable]`

### ISSUE THREAD

**Issue by shiftinv** [microsoft/pyright#2881]
**Describe the bug**
A condition using `isinstance(x, type)` with `x: Union[type, Callable[[], Any]]` narrows `x` to `type` in the true branch, and to `Never` in the false branch.

**Expected behavior**
The type to be `Callable[[], Any]`/`(() -> Any)`, not `Never`.

**Screenshots or Code**
```python
from typing import Any, Callable, Union

def f(x: Union[type, Callable[[], Any]]):
    reveal_type(x)         # Type of "x" is "type | (() -> Any)"
    if isinstance(x, type):
        reveal_type(x)     # Type of "x" is "type"
    else:
        reveal_type(x)     # Type of "x" is "Never"  <<
    reveal_type(x)         # Type of "x" is "type"   <<
```

**VS Code extension or command-line**
command-line, pyright 1.1.210

**Comment by erictraut** [comment]
Thanks for the bug report. I agree that this is a bug, but I don't think the code will do what you think it will.

The problem is that `Callable[[], Any]` might be an instance of `type`, so it's not safe for a type checker to assume that an `isinstance` can narrow the type to just `type` in this case.

Consider the following:
```python
from typing import Any, Callable

class MyClass:
    pass

def f(x: Callable[[], Any]):
    if isinstance(x, type):
        print("is type")

f(MyClass)  # Prints "is type"
```

**Comment by erictraut** [comment]
On second thought, it is safe to narrow to `type` in the positive ("if") case since that check will succeed only if it's `type` instance or an intersection of `type` and `Callable`. Either way, it's a `type`.

**Comment by shiftinv** [comment]
Thanks! That makes sense.

My particular use case for this just requires `x` to be a `type` inside the `if`, and handles other cases below. The `if isinstance(...)` seems to drop the `Callable` even outside of the true branch, which is what confused me:

```py
from typing import Any, Callable, Dict, Type, Union

class MyClass:
    pass

class MyClassA(MyClass):
    pass

lookup: Dict[Type[Any], Type[MyClass]] = {
    int: MyClassA,
    # ...
}

def f(x: Union[Type[Any], Callable[[str], Any]]):
    if isinstance(x, type):
        x = lookup.get(x, x)
        reveal_type(x)        # Type of "x" is "Type[MyClass] | type"
    reveal_type(x)            # Type of "x" is "Type[MyClass] | type"
```
The first one is correct, for the second one I would've expected `Type[MyClass] | type | Callable[[str], Any]` instead, or something along those lines (unless I'm missing something here).

**Comment by erictraut** [comment]
Yes, you're correct.

The `isinstance` narrowing needs to deal with the case where the class list passed as the second argument of `instance` includes a `@runtime_checkable` protocol class that defines a `__call__` method. In that case, it can narrow a `Callable` based on type compatibility of the callable signature. In your example, `type` does provide a `__call__` method, but it's not a `@runtime_checkable` protocol class, so it shouldn't have influenced the type narrowing in that manner.

This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.211, which I just published. It will also be included in the next release of pylance.
